### ISSUE TITLE
publish on F-Droid

### ISSUE THREAD

**Issue by rhjdvsgsgks** [dyhkwong/Exclave#105]
hi

Any plan to publish this application to F-Droid? Sagernet exists there before but was removed due to its discontinuation. Therefore, publishing a fork of Sagernet to F-Droid should not be too hard since upstream already meets the requirements. I guess the only thing missing is your consent to publish to F-Droid.

**Comment by klzgrad** [comment]
> should oppose redistribution

I don't oppose redistribution and the license permits it , but hopefully I have convinced you it is technically unnecessary and confusing for every plugin-integrating app to repackage it without substantial modification. I'm open to F-Droid packaging of these plugins for common use of SagerNet forks.

**Comment by linsui** [comment]
Data files are allowed and can even be non-free. But the geoip files should be from a legal source. We can use other geoip data. This is what we do in [NixOS](https://github.com/NixOS/nixpkgs/blob/master/pkgs/data/misc/v2ray-geoip/default.nix#L35). IIUC the maxmind data doesn't allow redistribution.

Is that difficult to generate the BrowserBridge js file?

Since SFA is reproducible I thought we can make Exclave reproducible, too.

**Comment by linsui** [comment]
> Is "browser bridge" considered to be a part of the code and need to be built from source?

Is it used? Not sure if this is needed for mobile.

Is the very old toolchain go 1.19.3? It's not very old for debian. :) https://packages.debian.org/bookworm/golang-go If we can't rebuild the officiale file you can rebuild it and it should be easier to be reproducible.

**Comment by linsui** [comment]
Ah, now I understand what it is. I thought it should be built from source. I feel that it's better to rebuild it with go 1.19 and update the hardcoded hash.

**Comment by linsui** [comment]
@dyhkwong I'm trying to packag exclave as reproducible build. Let's solve the problems one by one.

There are embeded paths and ndk build-id in the libtermux.so. The build id can be removed with `LOCAL_LDFLAGS := -Wl,--build-id=none`, the paths can be removed with `LOCAL_CFLAGS := -ffile-prefix-map=<NDK path>=ndk -ffile-prefix-map=<repo path>=Exclave`. Could you please add that?

**Comment by linsui** [comment]
Yep, it's reproducible now. :tada: Now let's solve the problem of geosite.dat and geoip.dat. The geosite.dat is totally find. But could you please upload the version.txt files to release so that I can get the version from them directly? The geoip.dat is problemic though. In fact v2ray itself and even the debian package violate the license of MaxMind GeoIP data. We can replace the MaxMind data with https://db-ip.com/db/download/ip-to-country-lite then run the generator on it.

**Comment by linsui** [comment]
This is what I use to pin the assets version

```
sed -i -e '/geoip/s/latestRelease/getReleaseByTagName("202412120057")/' -e '/domain-list-community/s/latestRelease/getReleaseByTagName("20241210004721")/' buildSrc/src/main/kotlin/V2RayAssets.kt
```

**Comment by linsui** [comment]
Looks like it's almost ready! Do you want to publish all 4 abis on F-Droid? F-Droid requires different apk has different version code, i.e., the version code of them should be something like 9331, 9332, 9333, 9334. Currently all of them have the same version code. And we need the metadata, as in https://github.com/SagerNet/SagerNet/tree/dev/metadata/en-US. Not sure if you want to also publish the plugins on F-Droid. It's better to use the Triple-T format so that we can add different metadata for the main app and the plugins. You can see an example in https://github.com/fcitx5-android/fcitx5-android/tree/master/app/src/main/play. See https://gitlab.com/-/snippets/1901490 for details.

**Comment by linsui** [comment]
Thanks!

**Comment by linsui** [comment]
IIUC it's all handled by the build scripts so it should work for F-Droid automatically? :)

**Comment by linsui** [comment]
Could you please make https://github.com/dyhkwong/LibSagerNetCore/blob/dec946cd1b7a2cc9a3ed7d613a3532aa29c3ad16/build.sh accept an arg to build a specific abi?

**Comment by linsui** [comment]
I just tested the latest commit and it works well.

**Comment by linsui** [comment]
You just need to release a new version. :)

**Comment by linsui** [comment]
Ah, I forgot the `DependencyInfoBlock` problem.

It's a [Signing block](https://bi-zone.medium.com/easter-egg-in-apk-files-what-is-frosting-f356aa9f4d1) added by AGP and encrypted with the Google public key so it can't be read by anyone else except Google. You can read more about it [here](https://gitlab.com/fdroid/admin/-/issues/367) and [here](https://gitlab.com/fdroid/fdroidserver/-/issues/1056).

Could you please disable it with the following code?

```
android {
    dependenciesInfo {
        // Disables dependency metadata when building APKs.
        includeInApk = false
        // Disables dependency metadata when building Android App Bundles.
        includeInBundle = false
    }
}
```

**Comment by linsui** [comment]
https://gitlab.com/fdroid/fdroiddata/-/merge_requests/17951 It will be built in the next cycle. It may take a week. Thanks for you help!

**Comment by linsui** [comment]
It has been published on https://f-droid.org/packages/com.github.dyhkwong.sagernet/ :tada:

**Comment by linsui** [comment]
The build failed. https://gitlab.com/fdroid/fdroiddata/-/jobs/8940347542 Could you please take a look?
